update geometry/triangulation (#102)
OI-wiki/i18n: docs/geometry/triangulation.md

![3 types of triangulation](./images/triangulation-0.svg)

The triangulation in OI mainly refers to the perfect triangulation in two-dimensional geometry (two-dimensional Delaunay triangulation, DT for abbreviation).

## Delaunay triangulation

### Definition

In mathematics and computational geometry, for a set of discrete points $P$ in a given plane, its Delaunay triangulation DT( $P$ ) satisfies:

1. Empty circumcircle: DT( $P$ ) is unique (any four points cannot be in a circle), in DT( $P$ ), there exists no other points in the circumcircle of **any** triangles.
2. Maximize the smallest angle: in the possible triangulations of the point set $P$ , the smallest angle of the triangle formed by DT( $P$ ) is the largest. In this sense, DT( $P$ ) is the triangulation that is **closest to regularization**. Specifically, a convex quadrilateral diagonals formed by two adjacent triangles. After exchange, the smallest angle of the two internal angles no longer increases.

![](./images/triangulation-1.png)

### Properties

1. Closest: a triangle is formed by the three closest points, and all line segments (sides of the triangle) do not intersect.

> 2. Uniqueness: no matter where the area is constructed, the final result will be consistent (any four points in the point set cannot be a circle).

Reviewer: ```suggestion
2. Uniqueness: no matter which point in the area is used as the start point of the construction, the final result will be consistent (any four points in the point set cannot be a circle).
```

3. Optimality: if the diagonals of a convex quadrilateral formed by any two adjacent triangles are interchangeable, the smallest angle among the six internal angles of the two triangles will not change.

> 4. The extreme rule: If the smallest angle of each triangle in the triangulation is arranged in ascending order, the Delaunay triangulation will get the largest value.

Reviewer: sort?

5. Regionality: adding, deleting, or moving any vertex will only affect the adjacent triangle.
6. Convex shell: the outermost boundary of the triangulation forms a convex polygonal shell.

## Construct a divide and conquer algorithm for DT

DT has many construction algorithms. Among the construction algorithms with time complexity of $O(n \log n)$ , the divide and conquer algorithm is the easiest to understand and implement.

> The first step of constructing DT by divide and conquer is to arrange the given point set according to the coordinates of $x$ in **ascending order**. The following figure shows the sorted point set of size $10$ .

Reviewer: ```suggestion
The first step of constructing DT by divide and conquer is to sort the given point set according to the coordinates of $x$ in **ascending order**. The following figure shows the sorted point set of size $10$ .
```

![Sorted point set of size $10$](./images/triangulation-2.svg)

> Once the point set is in order, we can continuously divide it into two parts (divide and conquer) until the size of the sub-point set does not exceed $3$ . Then these sub-point sets can be immediately divided into a triangle or line segments.

Reviewer: ordered

![分治为包含 2 或 3 个点的点集](./images/triangulation-3.svg)

> Then in the process of divide and conquer backtracking, the left and right sub-point sets that have been divided can be merged sequentially. The merged division contains LL-edge (the edge of the left sub-point set), RR-edge (the edge of the right sub-point set), and LR-edge (the new edge generated by connecting the left and right divisions), as shown in the figure LL-edge (gray), RR-edge (red), LR-edge (blue). For the merged subdivision, in order to maintain the nature of DT, we **may** need to delete some LL-edge and RR-edge, but we **will not** add LL-edge and RR-edge when merging.

Reviewer: ```suggestion
Then in the process of divide and conquer backtracking, the left and right sub-point sets that have been divided can be merged sequentially. The merged division contains LL-edge (the edge of the left sub-point set), RR-edge (the edge of the right sub-point set), and LR-edge (the new edge generated by connecting the left and right divisions), as shown in the figure LL-edge (gray), RR-edge (red), LR-edge (blue). For the merged subdivision, in order to maintain the properties of DT, we **may** need to delete some LL-edge and RR-edge, but we **will not** add LL-edge and RR-edge when merging.
```

![Edge](./images/triangulation-4.svg)

> The first step to merge the left and right splits is to insert the base LR-edge, which is the **bottom** LR-edge that does not intersect **any** LL-edge and RR-edge.

Reviewer: ```suggestion
The first step to merge the left and right splits is to insert the base LR-edge, which is the **bottom** LR-edge that does not intersects **any** LL-edge and RR-edge.
```

};
```

## Voronoi diagram

Voronoi 图由一组由连接两邻点直线的垂直平分线组成的连续多边形组成，根据 $n$ 个在平面上不重合种子点，把平面分成 $n$ 个区域，使得每个区域内的点到它所在区域的种子点的距离比到其它区域种子点的距离近。The Voronoi diagram is composed of a set of continuous polygons composed of vertical bisectors connecting two adjacent points. According to $n$ seed points that do not overlap on the plane, the plane is divided into $n$ regions, so that the points in each region The distance to the seed point in its area is shorter than the distance to the seed points in other areas.

> Voronoi 图是 Delaunay 三角剖分的对偶图，可以使用构造 Delaunay 三角剖分的分治算法求出三角网，再使用最左转线算法求出其对偶图实现在 $O(n \log n)$ 的时间复杂度下构造 Voronoi 图。

Reviewer: 中文

> The Voronoi diagram is the dual graph of Delaunay triangulation. You can use the divide and conquer algorithm to construct the Delaunay triangulation to find the triangulation, and then use the leftmost line algorithm to find the dual graph. It is implemented in $O(n \log n)$ for constructing the Voronoi diagram.

Author: 最左转线算法这个不太确定 :thinking:

Reviewer: 找到比较相关的就是[这个](https://en.wikipedia.org/wiki/Bowyer%E2%80%93Watson_algorithm) 但还没能建立平行关系

Reviewer: 相关：时间复杂度相等，V 的对偶图
